library-best-practices/JLBP-6.md

```diff
@@ -0,0 +1,165 @@
+- The Java package starts as `com.google.a1`
+  - With no rename, the new major version's Java package is still
+    `com.google.a1`.
+  - With a rename, the new major version's Java package is `com.google.a2`.
+
+Given this scenario, here are the possible combinations of renamings:
+
+- **Don't rename Java package**:
+  - **Don't rename Maven ID (case 1)**: This approach can result in diamond
+    dependency conflicts because different branches of a dependency tree can
+    depend on different major versions, and the build system (Maven or Gradle)
+    will only choose one of them to use. Users are forced to change their code
+    between versions, but only for library surface that was changed.
+  - **Rename Maven ID (case 2)**: Users can easily pull in both the old jar
+    (`g1:a1:1.0.0`) and the new jar (`g2:a2:2.0.0`) accidentally through
+    transitive dependencies, because Maven treats the two artifacts as distinct,
+    but since they have classes with the same Java packages, they are loaded in
+    unpredictable ways, leading to runtime exceptions (for example
+    `ClassNotFoundException`). It is difficult for users to ensure that their
+    build tree only includes one of the artifacts. **NEVER DO THIS**.
+- **Rename Java package**:
+  - **Don't rename Maven ID (case 3)**: The classes from `g1:a1:1.0.0` and
+    `g1:a1:2.0.0` could technically be used together, but since they share the
+    Maven group ID and artifact ID, only one jar can be pulled in. Users are
+    forced to change their code if they want to add any references to the new
+    version. It is strictly better to either also rename the Maven ID (case 4)
+    or to keep the Maven ID and also bundle the old package (case 5), so that
+    the major versions can be used side by side.
+  - **Rename Maven ID (case 4)**: The two major versions can be used side by
+    side, allowing users to incrementally transition from the old to the new
+    version, or even use them side by side indefinitely if necessary. This is
+    similar to the approach taken by Go, described in
+    https://research.swtch.com/vgo-import . The difference between Java and Go
+    is that Go can accept types with the same structure even though they have
+    different identities, so this approach is more disruptive in Java. In Go, a
+    single method can accept types from multiple interface declarations, but
+    Java can accept only one. Transitioning fully to the new version in Java
+    requires code changes between versions, even for classes whose surface
+    remains the same, because all import statements need to be updated. There
+    will be no diamond dependency conflicts using this approach, but adoption
+    can be blocked if there are consuming libraries that have not also created
+    new major versions that can accept new types from this library.
+  - **Bundle old and new in the existing Maven ID (case 5)**: Like case 4, the
```

Comment:
This may be a bit too specific. I am assuming this is when you have a shim from the old API to the new API. In that case I'd tend to suggest the old artifact (with old package) to depend on a new artifact (with new package). It does mean there are now two artifacts to choose from, but if you choose the old name things "just work."

Reply:
Actually I had in mind the junit case, where I'm guessing they don't use a shim. I had overlooked the shim idea. Do you feel it needs to be included?